parser-typechecker/src/Unison/Server/Backend.hs

```diff
@@ -912,61 +911,71 @@ resolveRootBranchHash mayRoot codebase = case mayRoot of
-  => NameScoping
-  -> Branch m
-  -> Codebase m v Ann
-  -> [HQ.HashQualified Name]
-  -> Backend m (DefinitionResults v)
-definitionsBySuffixes namesScope branch codebase query = do
-  -- First find the hashes by name and note any query misses.
-  QueryResult misses results <- lift
-    $ hqNameQuery namesScope branch codebase query
-  -- Now load the terms/types for those hashes.
-  results' <- lift $ loadSearchResults codebase results
-  let termTypes :: Map.Map Reference (Type v Ann)
-      termTypes = Map.fromList
-        [ (r, t) | SR'.Tm _ (Just t) (Referent.Ref r) _ <- results' ]
-      (collatedTypes, collatedTerms) = collateReferences
-        (mapMaybe SR'.tpReference results')
-        (mapMaybe SR'.tmReferent results')
-  -- load the `collatedTerms` and types into a Map Reference.Id Term/Type
-  -- for later
-  loadedDerivedTerms <-
-    lift $ fmap (Map.fromList . catMaybes) . for (toList collatedTerms) $ \case
-      Reference.DerivedId i -> fmap (i, ) <$> Codebase.getTerm codebase i
-      Reference.Builtin{}   -> pure Nothing
-  loadedDerivedTypes <-
-    lift $ fmap (Map.fromList . catMaybes) . for (toList collatedTypes) $ \case
-      Reference.DerivedId i ->
-        fmap (i, ) <$> Codebase.getTypeDeclaration codebase i
-      Reference.Builtin{} -> pure Nothing
-  -- Populate DisplayObjects for the search results, in anticipation of
-  -- rendering the definitions.
-  loadedDisplayTerms <- fmap Map.fromList . for (toList collatedTerms) $ \case
-    r@(Reference.DerivedId i) -> do
-      let tm = Map.lookup i loadedDerivedTerms
-      -- We add a type annotation to the term using if it doesn't
-      -- already have one that the user provided
-      pure . (r, ) $ case liftA2 (,) tm (Map.lookup r termTypes) of
-        Nothing        -> MissingObject $ Reference.idToShortHash i
-        Just (tm, typ) -> case tm of
-          Term.Ann' _ _ -> UserObject tm
-          _             -> UserObject (Term.ann (ABT.annotation tm) tm typ)
-    r@(Reference.Builtin _) -> pure $ (r,) $ case Map.lookup r B.termRefTypes of
-      Nothing -> MissingObject $ Reference.toShortHash r
-      Just typ -> BuiltinObject (mempty <$ typ)
-  let loadedDisplayTypes = Map.fromList . (`fmap` toList collatedTypes) $ \case
-        r@(Reference.DerivedId i) ->
-          (r, )
-            . maybe (MissingObject $ Reference.idToShortHash i) UserObject
-            $ Map.lookup i loadedDerivedTypes
-        r@(Reference.Builtin _) -> (r, BuiltinObject ())
-  pure $ DefinitionResults loadedDisplayTerms loadedDisplayTypes misses
+definitionsBySuffixes ::
+  forall m v.
+  MonadIO m =>
+  Var v =>
+  NameScoping ->
+  Branch m ->
+  Codebase m v Ann ->
+  IncludeCycles ->
+  [HQ.HashQualified Name] ->
+  m (DefinitionResults v)
+definitionsBySuffixes namesScope branch codebase includeCycles query = do
+  QueryResult misses results <- hqNameQuery namesScope branch codebase query
+  terms <- do
+    let termRefs0 = searchResultsToTermRefs results
```

Comment:
What does the zero mean here? Is there a more descriptive name like `termRefsWithoutCycles`?

Reply:
It just kinda means, "not the final version returned from this little `do`-block".

Reply:
I think that was my nice way of suggesting "can we find a better name here"? Or perhaps add a type or something to help distinguish it from `termRefs`?

Reply:
Ok, I renamed this to `termRefsWithoutCycles`, and `typeRefs0` to `typeRefsWithoutCycles` below